### PR TITLE
WIP: multilingual support.

### PR DESCRIPTION
**Pull request by cmungall**
# Multilanguage support

In the next few days the hpo international edition will be available from the sem-sql repo

This means you can do:

```
runoak -i sqlite:obo:hpinternational COMMAND ARGS
```

When this PR is merged and releases the commands below should all work with this edition.

## Caveats

 - Note that the precise logic for defaults is not yet settled. See examples below
 - May depend on some decisions in OMO, see https://github.com/information-artifact-ontology/ontology-metadata/issues/128
 - Only works with sql_implementation
     - support for direct rdf/owl coming soon
     - support for obo format indefinitely delayed depending on changes in 1.6

## Setting a non-default preferred language

`runoak --preferred-language fr -i tests/input/hp-international-test.db labels .ancestors HP:0020110`

|id|label|
|---|---|
|HP:0000001|Tous|
|HP:0000118|Anomalie phénotypique|
|HP:0000924|Anomalie du système squelettique|
|HP:0011842|Anomalie de la morphologie du squelette|
|HP:0020110|Bone fracture|
|HP:0033127|#Abnormalita muskuloskeletálního systému #Porucha pohybového aparátu|
|HP:0033127|Abnormality of the musculoskeletal system|

## Showing all languages

`runoak  -i tests/input/hp-international-test.db labels .ancestors HP:0020110 --all-languages`

|id|label|language|
|---|---|---|
|HP:0000001|All|None|
|HP:0000001|Alle|nl|
|HP:0000001|Tous|fr|
|HP:0000001|Tümü|tr|
|HP:0000001|Vše|cs|
|HP:0000118|Anomalie phénotypique|fr|
|HP:0000118|Fenotipik anormallik|tr|
|HP:0000118|Fenotypische abnormaliteit|nl|
|HP:0000118|Fenotypová abnormalita|cs|
|HP:0000118|Phenotypic abnormality|None|
|HP:0000924|Abnormalita kosterního systému|cs|
|HP:0000924|Abnormality of the skeletal system|None|
|HP:0000924|Afwijking van het skeletsysteem|nl|
|HP:0000924|Anomalie du système squelettique|fr|
|HP:0000924|İskelet sistemi anormalliği|tr|
|HP:0011842|Abnormal skeletal morphology|None|
|HP:0011842|Abnormalita morfologie skeletu|cs|
|HP:0011842|Afwijking van skelet morfologie|nl|
|HP:0011842|Anomalie de la morphologie du squelette|fr|
|HP:0011842|İskelet morfolojisi anormalliği|tr|
|HP:0020110|Bone fracture|None|
|HP:0033127|#Abnormalita muskuloskeletálního systému #Porucha pohybového aparátu|cs|
|HP:0033127|Abnormality of the musculoskeletal system|None|

## Pivoting by labels

`runoak  -i tests/input/hp-international-test.db labels .ancestors HP:0020110 --pivot-languages`

|id|None|nl|fr|tr|cs|
|---|---|---|---|---|---|
|HP:0000001|All|Alle|Tous|Tümü|Vše|
|HP:0000118|Phenotypic abnormality|Fenotypische abnormaliteit|Anomalie phénotypique|Fenotipik anormallik|Fenotypová abnormalita|
|HP:0000924|Abnormality of the skeletal system|Afwijking van het skeletsysteem|Anomalie du système squelettique|İskelet sistemi anormalliği|Abnormalita kosterního systému|
|HP:0011842|Abnormal skeletal morphology|Afwijking van skelet morfologie|Anomalie de la morphologie du squelette|İskelet morfolojisi anormalliği|Abnormalita morfologie skeletu|
|HP:0020110|Bone fracture|||||
|HP:0033127|Abnormality of the musculoskeletal system||||#Abnormalita muskuloskeletálního systému #Porucha pohybového aparátu|

TBD: should the column heading be None to truthfully relfect the underlying RDF (no lang) tag, or should `default_language` be induced?

## Code Examples

all examples are for BasicOntologyInterface

### parameterizing labels()

```python
>>> from oaklib import get_adapter
>>> adapter = get_adapter('tests/input/hp-international-test.db')
>>> adapter.label('HP:0000118', lang='fr')
'Anomalie phénotypique'
```

If the argument matches the default language, and the literal is not
explicitly language tagged, it is still returned

```python
>>> from oaklib import get_adapter
>>> adapter = get_adapter('tests/input/hp-international-test.db')
>>> adapter.label('HP:0000118', lang='en')
'Phenotypic abnormality'
```

### multilingual_labels method

```python
>>> from oaklib import get_adapter
>>> adapter = get_adapter('tests/input/hp-international-test.db')
>>> for tuple in sorted(adapter.multilingual_labels(['HP:0000118'], langs=['en', 'nl'])):
...     print(tuple)
('HP:0000118', 'Fenotypische abnormaliteit', 'nl')
('HP:0000118', 'Phenotypic abnormality', None)
```
Note that the english label is not explicitly tagged with a language tag, but is returned
because it matches the :ref:`default_language`.

### CODE REVIEW

File: src/oaklib/implementations/sparql/abstract_sparql_implementation.py [hunk 1]
```diff
@@ -388,7 +388,7 @@ def all_rdflib_triples(self) -> Iterator[RDF_TRIPLE]:
             raise NotImplementedError("Not implemented for SPARQL endpoints")
         yield from self.graph.triples((None, None, None))
 
-    def hierararchical_parents(self, curie: CURIE, isa_only: bool = False) -> List[CURIE]:
+    def hierarchical_parents(self, curie: CURIE, isa_only: bool = False) -> List[CURIE]:
```
Comment (matentzn) [inline src/oaklib/implementations/sparql/abstract_sparql_implementation.py:391]: As an aside: some more care is needed this kind of thing does not happen.. :P Sorry its unrelated. But I keep hearing from people like Anita that are relying heavily on these implementations that things change a lot from one version to another; I guess this is still ok right now but are using OAK more and more in production pipelines now.

**Review by matentzn — COMMENTED** [review]
Very nice!! If this is released before May we should include it in the HPO paper as a reference. We should also start writing an OAK paper soon!

File: src/oaklib/implementations/sparql/abstract_sparql_implementation.py (hunk 1 above)
Comment (cmungall) [reply src/oaklib/implementations/sparql/abstract_sparql_implementation.py:391]: understood. These should only occur between middle digit versions; this one is such a clear error it seemed reasonable to include this but there is an argument for upping the middle digit next release.

**Comment by cmungall** [comment]
yes this will definitely be released before May

**Pull request merged**